### ISSUE TITLE
The filtering behavior of List() and DeleteAllOf() are different when they are used with the "InNamespace" predicate.

### ISSUE THREAD

**Issue by zhranklin** [kubernetes-sigs/controller-runtime#988]
Hi, I'm using operator-sdk. When I'm using the k8s client provided by controller-manager constructed by operator-sdk, I found the following surprising things:

I am accidentally trying to delete CRs under a certain namespace, which is a mistake because the CR was defined in cluster scope. The strange thing is that DeleteAllOf() removes all CRs in the cluster, but List() didn't return any result.

The relevant code probably is:

cache_reader:

https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/cache/internal/cache_reader.go#L109

writer:

https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/unstructured_client.go#L130

**Comment by vincepri** [comment]
To clarify, you've called `DeleteAllOf` with a `InNamespace` predicate on a global custom resource?

**Comment by zhranklin** [comment]
@vincepri Yes. `DeleteAllOf` and `List` with a `InNamespace` on a global custom resource.

**Comment by vincepri** [comment]
Got it. We should definitely fix the behavior, returning an error when using `InNamespace` predicate on non-namespaced resources makes sense to me.
@alvaroaleman wdyt?

**Comment by alvaroaleman** [comment]
> Got it. We should definitely fix the behavior, returning an error when using InNamespace predicate on non-namespaced resources makes sense to me.

sounds good to me

**Comment by dlanov** [comment]
/assign

I verified that the namespace option is still silently ignored for cluster-scoped resources in the current client implementation. I’d like to reproduce the behavior with focused tests, then add the smallest validation needed so `InNamespace` returns an error for cluster-scoped resources.

I’ll check both `List` and `DeleteAllOf` and keep the change limited to the client packages.
